File: 22-Game-Theory/java/CatAndMouse.java
/*
 * Problem: Cat and Mouse
 * Source:  LeetCode 913 - https://leetcode.com/problems/cat-and-mouse/
 * 
 * Approach: BFS from terminal states, propagating game results backwards.
 * Time Complexity:  O(n^3)
 * Space Complexity: O(n^2)
 * 
 * Tags: game-theory, bfs, graph, minimax
 */

import java.util.*;

public class CatAndMouse {
    
    static final int DRAW = 0, MOUSE_WIN = 1, CAT_WIN = 2;
    
    public int catMouseGame(int[][] graph) {
        int n = graph.length;
        int[][][] result = new int[n][n][2];
        int[][][] degree = new int[n][n][2];
        
        for (int m = 0; m < n; m++) {
            for (int c = 0; c < n; c++) {
                degree[m][c][0] = graph[m].length;
                degree[m][c][1] = graph[c].length;
                for (int next : graph[c]) {
                    if (next == 0) degree[m][c][1]--;
                }
            }
        }
        
        Queue<int[]> queue = new LinkedList<>(); // {mouse, cat, turn, result}
        
        for (int c = 1; c < n; c++) {
            for (int t = 0; t < 2; t++) {
                result[0][c][t] = MOUSE_WIN;
                queue.offer(new int[]{0, c, t, MOUSE_WIN});
                result[c][c][t] = CAT_WIN;
                queue.offer(new int[]{c, c, t, CAT_WIN});
            }
        }
        
        while (!queue.isEmpty()) {
            int[] state = queue.poll();
            int m = state[0], c = state[1], t = state[2], r = state[3];
            int prevTurn = 1 - t;
            
            if (prevTurn == 0) { // mouse's previous turn
                for (int prevM : graph[m]) {
                    if (result[prevM][c][prevTurn] != DRAW) continue;
                    if (r == MOUSE_WIN) {
                        result[prevM][c][prevTurn] = MOUSE_WIN;
                        queue.offer(new int[]{prevM, c, prevTurn, MOUSE_WIN});
                    } else {
                        if (--degree[prevM][c][prevTurn] == 0) {
                            result[prevM][c][prevTurn] = CAT_WIN;
                            queue.offer(new int[]{prevM, c, prevTurn, CAT_WIN});
                        }
                    }
                }
            } else { // cat's previous turn
                for (int prevC : graph[c]) {
                    if (prevC == 0) continue;
                    if (result[m][prevC][prevTurn] != DRAW) continue;
                    if (r == CAT_WIN) {
                        result[m][prevC][prevTurn] = CAT_WIN;
                        queue.offer(new int[]{m, prevC, prevTurn, CAT_WIN});
                    } else {
                        if (--degree[m][prevC][prevTurn] == 0) {
                            result[m][prevC][prevTurn] = MOUSE_WIN;
                            queue.offer(new int[]{m, prevC, prevTurn, MOUSE_WIN});
                        }
                    }
                }
            }
        }
        
        return result[1][2][0];
    }
    
    public static void main(String[] args) {
        CatAndMouse sol = new CatAndMouse();
        int[][] graph = {{2,5},{3},{0,4,5},{1,4,5},{2,3},{0,2,3}};
        System.out.println("Result: " + sol.catMouseGame(graph));
    }
}

File: 22-Game-Theory/cpp/cat_and_mouse.cpp
/*
 * Problem: Cat and Mouse
 * Source:  LeetCode 913 - https://leetcode.com/problems/cat-and-mouse/
 * 
 * Approach: BFS from terminal states. State = (mouse_pos, cat_pos, whose_turn).
 *           Terminal states: mouse at 0 → mouse wins, cat at mouse → cat wins.
 *           Propagate results backwards using topological-sort-like BFS.
 * Time Complexity:  O(n^3) in worst case
 * Space Complexity: O(n^2)
 * 
 * Tags: game-theory, bfs, graph, minimax
 */

#include <bits/stdc++.h>
using namespace std;

const int MOUSE_WIN = 1, CAT_WIN = 2, DRAW = 0;
const int MOUSE_TURN = 0, CAT_TURN = 1;

int catMouseGame(vector<vector<int>>& graph) {
    int n = graph.size();
    // state[mouse][cat][turn]
    vector<vector<vector<int>>> result(n, vector<vector<int>>(n, vector<int>(2, DRAW)));
    vector<vector<vector<int>>> degree(n, vector<vector<int>>(n, vector<int>(2, 0)));
    
    // Initialize degrees
    for (int m = 0; m < n; m++) {
        for (int c = 0; c < n; c++) {
            degree[m][c][MOUSE_TURN] = graph[m].size();
            degree[m][c][CAT_TURN] = graph[c].size();
            // Cat cannot go to hole (node 0)
            for (int next : graph[c]) {
                if (next == 0) {
                    degree[m][c][CAT_TURN]--;
                }
            }
        }
    }
    
    queue<tuple<int,int,int,int>> q; // (mouse, cat, turn, result)
    
    // Terminal states
    for (int c = 1; c < n; c++) {
        for (int t = 0; t < 2; t++) {
            // Mouse at hole → mouse wins
            result[0][c][t] = MOUSE_WIN;
            q.push({0, c, t, MOUSE_WIN});
            // Cat catches mouse → cat wins
            result[c][c][t] = CAT_WIN;
            q.push({c, c, t, CAT_WIN});
        }
    }
    
    while (!q.empty()) {
        auto [m, c, t, r] = q.front(); q.pop();
        
        int prevTurn = 1 - t;
        
        if (prevTurn == MOUSE_TURN) {
            // Previous turn was mouse's — look at all states where mouse could have moved to m
            for (int prevM : graph[m]) {
                if (result[prevM][c][prevTurn] != DRAW) continue;
                
                if (r == MOUSE_WIN) {
                    // Mouse can choose this winning move
                    result[prevM][c][prevTurn] = MOUSE_WIN;
                    q.push({prevM, c, prevTurn, MOUSE_WIN});
                } else {
                    // r == CAT_WIN — mouse wouldn't choose this, but decrement degree
                    degree[prevM][c][prevTurn]--;
                    if (degree[prevM][c][prevTurn] == 0) {
                        result[prevM][c][prevTurn] = CAT_WIN;
                        q.push({prevM, c, prevTurn, CAT_WIN});
                    }
                }
            }
        } else {
            // Previous turn was cat's
            for (int prevC : graph[c]) {
                if (prevC == 0) continue; // cat can't be at hole
                if (result[m][prevC][prevTurn] != DRAW) continue;
                
                if (r == CAT_WIN) {
                    result[m][prevC][prevTurn] = CAT_WIN;
                    q.push({m, prevC, prevTurn, CAT_WIN});
                } else {
                    degree[m][prevC][prevTurn]--;
                    if (degree[m][prevC][prevTurn] == 0) {
                        result[m][prevC][prevTurn] = MOUSE_WIN;
                        q.push({m, prevC, prevTurn, MOUSE_WIN});
                    }
                }
            }
        }
    }
    
    return result[1][2][MOUSE_TURN]; // mouse at 1, cat at 2, mouse moves first
}

int main() {
    vector<vector<int>> graph = {{2,5},{3},{0,4,5},{1,4,5},{2,3},{0,2,3}};
    cout << "Result: " << catMouseGame(graph) << endl;
    // 0=Draw, 1=Mouse wins, 2=Cat wins
    
    return 0;
}
